### ISSUE TITLE
Neural Network: Constant Training Error when more than 40 Neurons per Layer

### ISSUE THREAD

**Issue by TobiasPicker** [accord-net/framework#1178]
**What would you like to submit?** (put an 'x' inside the bracket that applies)

  - [x] question
  - [ ] bug report
  - [ ] feature request

**Issue description**

Hello,

i'd like to train a neural network with 15 Inputs and 45 Outputs with the Accord.NET Framework in Unity3D. For that I have over 18.000 training data sets. Now i am trying to get the lowest training error possible with different network configurations. The training with the Parallel Resilient Backpropagation works fine, but when i try use more than 40 Neurons in a Layer the training error doesn't change anymore. It is stuck at the first error value at over 80.000. When I'm using the Identity function as the Activation Function it works with multiple hundreds of Neurons. What can cause the problem?

Tobias

**Comment by blaisexen** [comment]
@TobiasPicker 
Are you using this?
(http://accord-framework.net/docs/html/T_Accord_Neuro_Learning_ParallelResilientBackpropagationLearning.htm)

//The following example shows how to use Rprop to solve a multi-class classification problem.
// Suppose we would like to teach a network to recognize 
// the following input vectors into 3 possible classes:
// 
double[][] inputs =
{
    new double[] { 0, 1, 1, 0 }, // 0
    new double[] { 0, 1, 0, 0 }, // 0
    new double[] { 0, 0, 1, 0 }, // 0
    new double[] { 0, 1, 1, 0 }, // 0
    new double[] { 0, 1, 0, 0 }, // 0
    new double[] { 1, 0, 0, 0 }, // 1
    new double[] { 1, 0, 0, 0 }, // 1
    new double[] { 1, 0, 0, 1 }, // 1
    new double[] { 0, 0, 0, 1 }, // 1
    new double[] { 0, 0, 0, 1 }, // 1
    new double[] { 1, 1, 1, 1 }, // 2
    new double[] { 1, 0, 1, 1 }, // 2
    new double[] { 1, 1, 0, 1 }, // 2
    new double[] { 0, 1, 1, 1 }, // 2
    new double[] { 1, 1, 1, 1 }, // 2
};

int[] classes =
{
    0, 0, 0, 0, 0,
    1, 1, 1, 1, 1,
    2, 2, 2, 2, 2,
};

// First we have to convert this problem into a way that  the neural
// network can handle. The first step is to expand the classes into 
// indicator vectors, where a 1 into a position signifies that this
// position indicates the class the sample belongs to.
// 
double[][] outputs = Measures.Expand(classes, -1, +1);

// Create an activation function for the net
var function = new BipolarSigmoidFunction();

// Create an activation network with the function and
//  4 inputs, 5 hidden neurons and 3 possible outputs:
var network = new ActivationNetwork(function, 4, 5, 3);

// Randomly initialize the network
new NguyenWidrow(network).Randomize();

// Teach the network using parallel Rprop:
var teacher = new ParallelResilientBackpropagationLearning(network);

double error = 1.0;
while (error > 1e-5)
    error = teacher.RunEpoch(inputs, outputs);


// Checks if the network has learned
for (int i = 0; i < inputs.Length; i++)
{
    double[] answer = network.Compute(inputs[i]);

    int expected = classes[i];
    int actual; answer.Max(out actual);

    // actual should be equal to expected
}

I have open an issue here, do you think you can help me with some ideas why?
https://github.com/accord-net/framework/issues/1184

**Comment by TobiasPicker** [comment]
@blaisexen I'm not using it for classification, but my Code is mostly the same. The results with smaller networks are ok. I don't think that my problem is caused by the Learning Algorithm, it's the sigmoid activation function that behaves strange.

**Comment by blaisexen** [comment]
hi,
have you look unto my code?
any comment you think?
https://github.com/accord-net/framework/issues/1184

**Comment by TobiasPicker** [comment]
@blaisexen I can't find any mistakes in  your code. But like i said, i never did classification with neural networks. So i have no experience with your kind of problem.

**Comment by blaisexen** [comment]
Ok,
thank you, I think I got it now, latest test shows no false, all are correct.

**Comment by TobiasPicker** [comment]
@blaisexen I found my problem thanks to your code! I was missing this line: `new NguyenWidrow(network).Randomize();` in my Code. With the normal weight initialization the error was extremly high and didn't change in bigger networks. Now with the NguyenWidrow it gets better results in less time and it works with big networks too. 
Thank you for your help!

**Comment by blaisexen** [comment]
hi,
Glad to know that, but have you tried using multi-class classification then get top 3 maximum results then training the network to calculate XOR function?

Every top 3 result of the classification will have training again.

I mean is that an image recognition your working for? or just a test results like the samples given here?
https://github.com/accord-net/framework/tree/development/Samples/Neuro/Resilient%20Backpropagation

As of now there is no error result in my application, 100% all correct, but I want to classify the top 3 results and use it to calculate XOR function and get the 1000% sureness.

**Comment by TobiasPicker** [comment]
I'm doing a Motion Estimation from reduced Motion Capture Data. For this i am converting the output values to positions. So i have no classification. My project works fine, but i am trying to get the error value lower so that the estimation will be more accurate.

**Comment by blaisexen** [comment]
Motion Estimation from reduced Motion Capture? Cool, interesting! :dagger: 

thanks for the time :dango:

**Comment by TobiasPicker** [comment]
It definitely is 🤣  It's my Bachelor thesis 😁 
Thank you, too!
